module_test.go

```diff
@@ -420,4 +612,122 @@ func TestModuleFailures(t *testing.T) {
 			})
 		}
 	})
+
+	t.Run("invalid WithLogger in Module", func(t *testing.T) {
+		t.Parallel()
+
+		var spy fxlog.Spy
+
+		spyAsLogger := []fx.Option{
+			fx.Supply(&spy),
+			fx.WithLogger(func(spy *fxlog.Spy) fxevent.Logger {
+				return spy
+			}),
+		}
+
+		tests := []struct {
+			desc            string
+			giveModuleOpts  []fx.Option
+			giveAppOpts     []fx.Option
+			wantErrContains []string
+			wantEventTypes  []string
+		}{
+			{
+				desc:           "error in Provide shows logs in module",
+				giveModuleOpts: append(spyAsLogger, fx.Provide(&bytes.Buffer{})), // not passing in a constructor
+				giveAppOpts:    []fx.Option{},
+				wantErrContains: []string{
+					"must provide constructor function, got  (type *bytes.Buffer)",
+				},
+				wantEventTypes: []string{
+					"Supplied", "Provided", "LoggerInitialized",
+				},
+			},
+			{
+				desc: "logger in module failed to build",
+				giveModuleOpts: []fx.Option{
+					fx.WithLogger(func() (fxevent.Logger, error) {
+						return nil, errors.New("error building logger")
+					}),
+				},
+				giveAppOpts:     spyAsLogger,
+				wantErrContains: []string{"error building logger"},
+				wantEventTypes: []string{
+					"Supplied", "Provided", "Provided", "Provided",
+					"LoggerInitialized", "Provided", "LoggerInitialized",
+				},
+			},
+			{
+				desc: "logger dependency in module failed to build",
+				giveModuleOpts: []fx.Option{
+					fx.Provide(func() (*zap.Logger, error) {
+						return nil, errors.New("error building logger dependency")
+					}),
+					fx.WithLogger(func(log *zap.Logger) fxevent.Logger {
+						t.Errorf("WithLogger must not be called")
+						panic("must not be called")
+					}),
+				},
+				giveAppOpts:     spyAsLogger,
+				wantErrContains: []string{"error building logger dependency"},
+				wantEventTypes: []string{
+					"Supplied", "Provided", "Provided", "Provided",
+					"LoggerInitialized", "Provided", "Provided", "LoggerInitialized",
+				},
+			},
+			{
+				desc: "Invalid input for WithLogger",
+				giveModuleOpts: []fx.Option{
+					fx.WithLogger(&fxlog.Spy{}), // not passing in a constructor for WithLogger
+				},
+				giveAppOpts: spyAsLogger,
+				wantErrContains: []string{
+					"fx.WithLogger", "from:", "Failed",
+				},
+				wantEventTypes: []string{
+					"Supplied", "Provided", "Provided", "Provided",
+					"LoggerInitialized", "Provided", "LoggerInitialized",
+				},
+			},
+		}
+		for _, tt := range tests {
+			spy.Reset()
+			t.Run(tt.desc, func(t *testing.T) {
+				redis := fx.Module("redis",
+					append(
+						tt.giveModuleOpts,
+						fx.Provide(func() string {
+							return "redis"
+						}),
+						fx.Invoke(func(s string) {
+							assert.Fail(t, "this should never run")
+						}),
+					)...,
```

Comment:
collect the default options into `fx.Options` instead of appending and exploding it like this.